grafana/grafana
# Alerting: Fix panics when attempting to create an Alertmanager after failing

### Description

This PR updates the `grafana/alerting` and `grafana/prometheus-alertmanager` packages to include fixes to panics after failing to create an Alertmanager for an org and retrying (duplicate metrics collector registration).

Changes included in the `grafana/alerting` update:
- https://github.com/grafana/alerting/pull/249
- https://github.com/grafana/alerting/pull/250

Changes included in the `grafana/prometheus-alertmanager` update:
- https://github.com/grafana/prometheus-alertmanager/pull/90
- https://github.com/grafana/prometheus-alertmanager/pull/91
- https://github.com/grafana/prometheus-alertmanager/pull/92

_Fixes https://github.com/grafana/grafana/issues/92707_

### How to test it

After pulling the branch, return an error from [NewAlertmanager()](https://github.com/grafana/grafana/blob/7bb3fe3da1339ece7d4ce3dd79ebbe05ca0321f0/pkg/services/ngalert/notifier/alertmanager.go#L92), example:

```go
func NewAlertmanager(ctx context.Context, orgID int64, cfg *setting.Cfg, store AlertingStore, stateStore stateStore,
	peer alertingNotify.ClusterPeer, decryptFn alertingNotify.GetDecryptedValueFn, ns notifications.Service,
	m *metrics.Alertmanager, withAutogen bool,
) (*alertmanager, error) {
    return nil, fmt.Errorf("test error")
```

Run `make run`, the Alertmanager should fail to be created repeatedly over time, but you should get no panics. Attempting to do the same in `main` will result in Grafana crashing.
